File: src/main/java/com/ternbusty/takoyaki/util/json/JsonWriter.java
package com.ternbusty.takoyaki.util.json;

import java.util.List;
import java.util.Map;

/**
 * Writes a {@link JsonParser}-shaped tree (Map, List, String, Long, Double,
 * Boolean, null) back as JSON text.
 *
 * <p>Pretty-prints by default with 2-space indent — the OCI conformance
 * suite parses both formats, but pretty output is human-readable and
 * matches what jackson's {@code writerWithDefaultPrettyPrinter} produced
 * for our state.json files. {@link #toCompact} is available for places
 * we want one-line output (currently unused; OCI doesn't care).
 *
 * <p>Encoding rules: 32-bit and 64-bit integers go through as JSON numbers
 * without decimal points; doubles round-trip via {@link Double#toString}.
 * Strings escape only the characters strict JSON forbids unescaped (control
 * chars + {@code "} + {@code \\}); we do not pre-escape forward slash.
 */
public final class JsonWriter {
    private static final String INDENT = "  ";

    private JsonWriter() {}

    public static String toPretty(Object node) {
        StringBuilder sb = new StringBuilder();
        writeValue(sb, node, 0, true);
        return sb.toString();
    }

    public static String toCompact(Object node) {
        StringBuilder sb = new StringBuilder();
        writeValue(sb, node, 0, false);
        return sb.toString();
    }

    @SuppressWarnings("unchecked")
    private static void writeValue(StringBuilder sb, Object v, int depth, boolean pretty) {
        if (v == null) { sb.append("null"); return; }
        if (v instanceof Boolean b) { sb.append(b ? "true" : "false"); return; }
        if (v instanceof String s) { writeString(sb, s); return; }
        if (v instanceof Number n) { writeNumber(sb, n); return; }
        if (v instanceof Map<?, ?> m) { writeObject(sb, (Map<String, Object>) m, depth, pretty); return; }
        if (v instanceof List<?> l) { writeArray(sb, (List<Object>) l, depth, pretty); return; }
        throw new IllegalArgumentException("json write: unsupported type " + v.getClass());
    }

    private static void writeObject(StringBuilder sb, Map<String, Object> m, int depth, boolean pretty) {
        if (m.isEmpty()) { sb.append("{ }"); return; }
        sb.append('{');
        boolean first = true;
        for (Map.Entry<String, Object> e : m.entrySet()) {
            if (!first) sb.append(',');
            first = false;
            if (pretty) { sb.append('\n'); indent(sb, depth + 1); }
            writeString(sb, e.getKey());
            sb.append(pretty ? " : " : ":");
            writeValue(sb, e.getValue(), depth + 1, pretty);
        }
        if (pretty) { sb.append('\n'); indent(sb, depth); }
        sb.append('}');
    }

    private static void writeArray(StringBuilder sb, List<Object> a, int depth, boolean pretty) {
        if (a.isEmpty()) { sb.append("[ ]"); return; }
        sb.append('[');
        boolean first = true;
        for (Object e : a) {
            if (!first) sb.append(',');
            first = false;
            if (pretty) { sb.append('\n'); indent(sb, depth + 1); }
            writeValue(sb, e, depth + 1, pretty);
        }
        if (pretty) { sb.append('\n'); indent(sb, depth); }
        sb.append(']');
    }

    private static void writeString(StringBuilder sb, String s) {
        sb.append('"');
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c) {
                case '"' -> sb.append("\\\"");
                case '\\' -> sb.append("\\\\");
                case '\n' -> sb.append("\\n");
                case '\r' -> sb.append("\\r");
                case '\t' -> sb.append("\\t");
                case '\b' -> sb.append("\\b");
                case '\f' -> sb.append("\\f");
                default -> {
                    if (c < 0x20) {
                        sb.append(String.format("\\u%04x", (int) c));
                    } else {
                        sb.append(c);
                    }
                }
            }
        }
        sb.append('"');
    }

    private static void writeNumber(StringBuilder sb, Number n) {
        if (n instanceof Long || n instanceof Integer || n instanceof Short || n instanceof Byte) {
            sb.append(n.longValue());
        } else if (n instanceof Double || n instanceof Float) {
            double d = n.doubleValue();
            if (Double.isNaN(d) || Double.isInfinite(d)) {
                throw new IllegalArgumentException("json write: non-finite number " + d);
            }
            // Prefer integer form when possible — keeps state.json clean
            // (`pid: 4242` not `pid: 4242.0`).
            if (d == Math.rint(d) && Math.abs(d) < 1e15) {
                sb.append((long) d);
            } else {
                sb.append(d);
            }
        } else {
            sb.append(n);
        }
    }

    private static void indent(StringBuilder sb, int depth) {
        for (int i = 0; i < depth; i++) sb.append(INDENT);
    }
}

File: src/main/java/com/ternbusty/takoyaki/util/json/JsonWriter.kt
package com.ternbusty.takoyaki.util.json

/**
 * Writes a [JsonParser]-shaped tree (Map, List, String, Long, Double,
 * Boolean, null) back as JSON text.
 *
 * Pretty-prints by default with 2-space indent — the OCI conformance
 * suite parses both formats, but pretty output is human-readable and
 * matches what jackson's `writerWithDefaultPrettyPrinter` produced
 * for our state.json files. [toCompact] is available for places
 * we want one-line output (currently unused; OCI doesn't care).
 *
 * Encoding rules: 32-bit and 64-bit integers go through as JSON numbers
 * without decimal points; doubles round-trip via [Double.toString].
 * Strings escape only the characters strict JSON forbids unescaped (control
 * chars + `"` + `\`); we do not pre-escape forward slash.
 */
object JsonWriter {
    private const val INDENT = "  "

    fun toPretty(node: Any?): String {
        val sb = StringBuilder()
        writeValue(sb, node, 0, true)
        return sb.toString()
    }

    fun toCompact(node: Any?): String {
        val sb = StringBuilder()
        writeValue(sb, node, 0, false)
        return sb.toString()
    }

    @Suppress("UNCHECKED_CAST")
    private fun writeValue(sb: StringBuilder, v: Any?, depth: Int, pretty: Boolean) {
        when {
            v == null -> sb.append("null")
            v is Boolean -> sb.append(if (v) "true" else "false")
            v is String -> writeString(sb, v)
            v is Number -> writeNumber(sb, v)
            v is Map<*, *> -> writeObject(sb, v as Map<String, Any?>, depth, pretty)
            v is List<*> -> writeArray(sb, v as List<Any?>, depth, pretty)
            else -> throw IllegalArgumentException("json write: unsupported type ${v.javaClass}")
        }
    }

    private fun writeObject(sb: StringBuilder, m: Map<String, Any?>, depth: Int, pretty: Boolean) {
        if (m.isEmpty()) { sb.append("{ }"); return }
        sb.append('{')
        var first = true
        for ((key, value) in m) {
            if (!first) sb.append(',')
            first = false
            if (pretty) { sb.append('\n'); indent(sb, depth + 1) }
            writeString(sb, key)
            sb.append(if (pretty) " : " else ":")
            writeValue(sb, value, depth + 1, pretty)
        }
        if (pretty) { sb.append('\n'); indent(sb, depth) }
        sb.append('}')
    }

    private fun writeArray(sb: StringBuilder, a: List<Any?>, depth: Int, pretty: Boolean) {
        if (a.isEmpty()) { sb.append("[ ]"); return }
        sb.append('[')
        var first = true
        for (e in a) {
            if (!first) sb.append(',')
            first = false
            if (pretty) { sb.append('\n'); indent(sb, depth + 1) }
            writeValue(sb, e, depth + 1, pretty)
        }
        if (pretty) { sb.append('\n'); indent(sb, depth) }
        sb.append(']')
    }

    private fun writeString(sb: StringBuilder, s: String) {
        sb.append('"')
        for (c in s) {
            when (c) {
                '"' -> sb.append("\\\"")
                '\\' -> sb.append("\\\\")
                '\n' -> sb.append("\\n")
                '\r' -> sb.append("\\r")
                '\t' -> sb.append("\\t")
                '\b' -> sb.append("\\b")
                '\u000C' -> sb.append("\\f")
                else -> {
                    if (c < ' ') {
                        sb.append("\\u%04x".format(c.code))
                    } else {
                        sb.append(c)
                    }
                }
            }
        }
        sb.append('"')
    }

    private fun writeNumber(sb: StringBuilder, n: Number) {
        when (n) {
            is Long, is Int, is Short, is Byte -> sb.append(n.toLong())
            is Double, is Float -> {
                val d = n.toDouble()
                if (d.isNaN() || d.isInfinite()) {
                    throw IllegalArgumentException("json write: non-finite number $d")
                }
                // Prefer integer form when possible — keeps state.json clean
                // (`pid: 4242` not `pid: 4242.0`).
                if (d == Math.rint(d) && Math.abs(d) < 1e15) {
                    sb.append(d.toLong())
                } else {
                    sb.append(d)
                }
            }
            else -> sb.append(n)
        }
    }

    private fun indent(sb: StringBuilder, depth: Int) {
        repeat(depth) { sb.append(INDENT) }
    }
}
